packages/@atjson/document/src/query.ts

```diff
@@ -67,114 +103,259 @@ function set(object: any, key: string, value: any) {
     }
     set(object[path], rest.join('.'), value);
   }
-  return;
 }
 
-function flattenPropertyPaths(mapping: Renaming, options: { keys: boolean, values?: boolean }, prefix?: string): FlattenedRenaming {
-  return Object.keys(mapping).reduce((result: Renaming, key: string) => {
-    let value = mapping[key];
-    let fullyQualifiedKey = key;
-    if (prefix) {
-      fullyQualifiedKey = `${prefix}.${key}`;
-      if (options.values) {
-        value = `${prefix}.${value}`;
-      }
-    }
-    if (typeof value !== 'object') {
-      result[fullyQualifiedKey] = value;
-    } else {
-      Object.assign(result, flattenPropertyPaths(value, options, fullyQualifiedKey));
-    }
-    return result;
-  }, {});
+export class StrictMatch {
+  [key: string]: any;
+}
+
+export type JoinableAnnotation = Annotation | AnnotationJoin;
+
+export type FilterFunction = (annotation: JoinableAnnotation) => boolean;
+export type JoinFilterFunction = (leftAnnotation: JoinableAnnotation, rightAnnotation: JoinableAnnotation) => boolean;
+
+export class AnnotationJoin {
+
+  // Advice sought to make this not shit.
+  [key: string]: JoinableAnnotation | JoinableAnnotation[] | any;
+
+  constructor(left: JoinableAnnotation, name: string) {
+    this[name] = left;
+  }
+
+  addJoin(right: JoinableAnnotation[], name: string) {
+    this[name] = right;
+  }
 }
 
-function matches(annotation: any, filter: Filter): boolean {
+function matches(annotation: object, filter: StrictMatch): boolean {
   return Object.keys(filter).every(key => {
     let value = filter[key];
     if (typeof value === 'object') {
-      return matches(annotation[key], value);
+      return matches((annotation as any)[key], value);
     }
-    return annotation[key] === value;
+    return (annotation as any)[key] === value;
   });
 }
 
-export default class Query {
-  filter: Filter;
-  private document: Document;
-  private transforms: Transform[];
-  private currentAnnotations: Annotation[];
+export default class AnnotationCollection {
+  document: Document;
+  annotations: JoinableAnnotation[];
+  name?: string;
 
-  constructor(document: Document, filter: Filter) {
+  constructor(document: Document, annotations: JoinableAnnotation[] = []) {
     this.document = document;
-    this.filter = filter;
-    this.transforms = [(annotation: Annotation) => annotation];
-    this.currentAnnotations = document.annotations.filter(annotation => matches(annotation, this.filter));
+    this.annotations = annotations;
   }
 
-  run(newAnnotation: Annotation): Annotation[] {
-    // Release the list of currently filtered annotations
-    this.currentAnnotations = [];
-    if (matches(newAnnotation, this.filter)) {
-      let alteredAnnotations = this.transforms.reduce((annotations: Annotation[], transform: Transform) => {
-        return flatten(annotations.map(transform));
-      }, [newAnnotation]);
+  where(filter: StrictMatch | FilterFunction): AnnotationCollection {
```

Comment:
Could you add `[Symbol.iterator]` so this can be iterated over using `for...of`?

cf:

```js
*[Symbol.iterator]() {
  this.annotations.forEach((annotation) => yield annotation);
}
```